### PR TITLE
Enhance customization options

### PR DESCRIPTION
**Pull request by scott-8**
@shapiromatron

### CODE REVIEW

**Comment by J535D165** [comment]
I like the idea. If this is PR gets merged, I think we should consider changing the behavior of the `strict` argument as well. If strict is `False`, it can determine whether it is a list type or not on the fly. Just like ignoring tags if they are not recognized.

**Comment by holub008** [comment]
+1, code changes look reasonable.

Will let maintainers comment on if this expands the API too much - theoretically we could avoid any added `load`/`loads` arguments by requiring callers to supply their own `RisImplementation` with custom list tags. But that could be a bit clumsy.

**Comment by scott-8** [comment]
I like the idea of being able to customize `RisImplementation` a lot, it would work really well for me. The `mapping` parameter could even be moved there.

@J535D165 Do you think we'd need to add another parameter to control that? Right now, all `strict` does is determine the behavior for lines that don't have tags (skip or use previous line's tag). I could see someone wanting to skip lines with missing tags and wanting to enforce the list type tags at the same time, and one parameter wouldn't work in that scenario. It'd be a good idea to have a discussion about this, because I think `strict` can be improved.

**Comment by J535D165** [comment]
I think I was mistaken about the `strict` argument. `strict` might be confusing for users. 

I really like the idea of custom `RisImplementation`s. I think this is the way forward. The first steps towards this were made in this PR https://github.com/MrTango/rispy/pull/18#discussion_r382919192. I think this might be the moment to move this forward.  This makes it so much easier to write your own RIS parsers (and contribute them back to the project). 

```
from rispy import Ris

class MyCustomRisParser(Ris):
    def __init__(self, list_tags=["KW", "N1"], *args, **kwargs):
        super(MyCustomRisParser, self).__init__(list_tags = list_tags, *args, **kwargs)
```

```
import rispy

rispy.load("myfile.ris", implementation=MyCustomRisParser)
```

We can even try to simplify it further, however, I'm not sure about that. 


```
from rispy import Ris

class MyCustomRisParser(Ris):
   list_tags = ["KW", "N1"]
```

```
import rispy

rispy.load("myfile.ris", implementation=MyCustomRisParser)
```

Implementations that are officially supported can be called by their handle, like "wok".

```
import rispy

rispy.load("myfile.ris", implementation="wok")
```

but also with 

```
import rispy
from rispy import Wok

rispy.load("myfile.ris", implementation=Wok)
```

 Same interface for the mapping. 

Ideally, we have an option `implementation="auto"` one day :)

I think @scott-8's implementation is almost ready for the interface described!

**Comment by holub008** [comment]
Relying on `RisImplementation` feels most correct for the API that's been built here. However, I suspect the `RisImplementation` will rarely ever be overriden with anything more than data changes (i.e. the behavior of its methods will rarely change); when they're only overriding data, it feels inappropriate to have the caller deal with implementation details & polymorphism. To your point,  defaults/auto may solve this problem. 

If we go for the route of all args being passed through `RISImplementation`, we'll probably want to remove `strict` from `load` params as well.

RE `strict`: my team added that in https://github.com/MrTango/rispy/issues/30 to support RIS exports from several large databases that plague their users with non-standard formats. I'd hesitate to extend its function, especially when it comes to dynamically typing outputs from `loads`.

**Comment by scott-8** [comment]
I totally agree with your point @J535D165, I think this is a great step to make to make it easy to contribute parsers for more data sources.

I don't think we should need to create an entire new custom class if all we want to do is change tags. An idea of what this could look like:
```python
from rispy import Ris

MyCustomTags = Ris(list_tags=["N1", "KW"])
rispy.load("myfile.ris", implementation=MyCustomTags)
```

Then if you had some sort of custom format you could create a new parser like this:
```python
from rispy import Base

class MyNewFormat(Base):
    START_TAG = "TY"
    IGNORE = ["FN", "VR", "EF"]
    PATTERN = r"^[A-Z][A-Z0-9] |^ER\s?|^EF\s?"
    default_mapping = TAG_KEY_MAPPING
    default_list_tags = ["N1", "KW"]

    def get_content(self, line):
        return True

rispy.load("myfile.ris", implementation=MyNewFormat)
```
If someone wanted to contribute back, all they would have to do is add `MyNewFormat`, and nothing should need to be overridden for most cases. Custom formats could inherit `Base` or another existing format, depending on how much needs changing.

@holub008 I agree with removing `strict` from the load params. It should probably be added as a class variable that can be changed with a custom parser.

**Comment by shapiromatron** [comment]
I like the idea of a class-based implementation too. We never minted a 1.0 release so we're free to break the API. @scott-8 I like your suggestion here; what do others think? @J535D165 @holub008

**Comment by holub008** [comment]
@scott-8 Agreed that the RIS implementation should be an instantiation, not a subclass of `Ris`, for both varying `strict` and `list_tags` params. 

@shapiromatron My only hesitation is that the`Base` interface is fairly tightly coupled to the parsing algo, to the point that clients aren't likely to override most of its methods. So it's exposing unnecessary implementation internals. 

But I also don't think it's a big deal. Our user base is presumably modest and our target functionality is well defined. Providing control over `mapping`, `strict` and `list_tags` covers 99% of variation clients will need. Having 3 function params for these is easier and exposes less in the API, but we may run into issues for the 1% of cases where more flexibility is needed. I'm not sure users will hold it against us if they have to do:

```python
my_parser = Ris(mapping, strict, list_tags)
rispy.parse(data, my_parser)
````

vs
```python
rispy.parse(data, mapping, strict, list_tags)
```

to achieve default overrides.

**Comment by scott-8** [comment]
In my opinion, being able to create a custom parser is important because of the wide variety of RIS files. For example, say someone found an RIS file where there's only 1 space after each tag instead of 2. Without a custom parser, this package would not be suitable for parsing the file. While not many people may encounter something like this, it would be nice to have an easy way to solve such a problem instead of forcing them to fork the project.

I do agree about exposing unnecessary internals, but it personally doesn't seem that detrimental. We could add an underscore to the methods to suggest that people shouldn't be using them. We could also create some sort of custom object that contains all of the configurable parameters, and pass that to `parse` instead of a subclass of `Base`. It's not as elegant internally, however.

```python

class BaseConfig:
    START_TAG: str = None
    END_TAG: str = "ER"
    IGNORE: List[str] = []
    PATTERN: str = None

    def __init__(self, mapping, list_tags):
        self.mapping = mapping
        self.list_tags = list_tags

    @classmethod
    def get_content(cls, line):
        raise NotImplementedError

    @classmethod
    def is_counter(cls, line):
        return True

class Ris(BaseConfig):
    START_TAG = "TY"
    PATTERN = r"^[A-Z][A-Z0-9]  - |^ER  -\s*$"
    default_mapping = TAG_KEY_MAPPING
    default_list_tags = LIST_TYPE_TAGS

    counter_re = re.compile("^[0-9]+.")

    @classmethod
    def get_content(cls, line):
        return line[6:].strip()

    @classmethod
    def is_counter(cls, line):
        none_or_match = cls.counter_re.match(line)
        return bool(none_or_match)

NewTags = Ris(list_tags=["KW"])

```

Then you can use an instance of `Ris` if you want to change the tags, or create a new parser based on `BaseConfig`. It would then be passed to `parse`. This wouldn't expose any internals, but I don't like it as much.

**Comment by scott-8** [comment]
@shapiromatron Pushed a commit changing customization to a class-based implementation. Right now, it fails the tests because of the readme. Everything else should be fine, though.

Usage should look like was written above.
```python
from rispy import Ris

MyCustomTags = Ris(list_tags=["N1", "KW"])
rispy.load("myfile.ris", implementation=MyCustomTags)
```
```python
from rispy import Ris

class MyCustomRisParser(Ris):
    START_TAG = "TT"

rispy.load("myfile.ris", implementation=MyCustomRisParser())
```

Right now, the parser class takes `mapping` and `list_tags`. The writer class takes the same, with the addition of `type_of_reference` (wasn't sure what people used this for so I kept it).

I converted the `strict` parameter to a class variable and added a few more customization variables. I put these as class variables rather than instance variables since changing them will have a large effect on how something is parsed.
* `SKIP_MISSING_TAGS`: Same as `strict` in reverse, will skip lines with no tags (only in parser)
* `SKIP_UNKNOWN_TAGS`: Won't read or write tags not in the tag map
* `ENFORCE_LIST_TAGS`: Per @J535D165's suggestion, turning this `False` will automatically cause data to be parsed as list tags if there's more than one tag with that name. The list tag type list will still be used (i.e. for tags that only have one value). For writing, this will make everything be written as multiple tags if possible (normally lists are written as strings in one tag if not a list type tag)

Let me know what you think of these changes. I've held off on changing the Readme before we decide to move forward.

File: rispy/parser.py
Comment (shapiromatron) [inline rispy/parser.py:36]: these should be required; with no default

File: rispy/parser.py
Comment (shapiromatron) [inline rispy/parser.py:184]: rename to `clean_text`?

File: rispy/parser.py
Comment (shapiromatron) [inline rispy/parser.py:243]: change to Optional[BaseParser]?

File: rispy/parser.py
Comment (shapiromatron) [inline rispy/parser.py:284]: suggest breaking API here (and documenting in PR) and passing a `Optional[BaseParser]` with a default of None if None, then the load could instantiate a `RisParser()` class.

File: rispy/writer.py
Comment (shapiromatron) [inline rispy/writer.py:20]: same comments in the parser file would apply here. Remove `WriterImplemenation, make mappings required instead of optional in BaseWriter, move mapping property to init, etc

File: rispy/writer.py
Comment (shapiromatron) [inline rispy/writer.py:150]: same api changes to dump/dumps as suggested in load/loads

**Review by shapiromatron — CHANGES_REQUESTED** [review]
Added a few recommendations for changes, lots of comments but generally pretty minimal I think. Looking really nice!

File: rispy/parser.py
Comment (scott-8) [reply rispy/parser.py:202]: I can change the docstring, but I think this should be kept as something that relates the fact that this method will only operate on the header, compared to `SKIP_MISSING_TAGS`

File: rispy/parser.py
Comment (scott-8) [reply rispy/parser.py:15]: I think this is a good idea. The reason I used a string enum was to allow people to choose from a set of default implementations without separately importing them, but it's a little complicated.

File: rispy/parser.py
Comment (shapiromatron) [reply rispy/parser.py:202]: makes sense, yeah I'd just change it so the method name and docstring match and describes the purpose a little more clearly

**Comment by shapiromatron** [comment]
@scott-8 is making some breaking changes here based on the discussion above but I think they are good improvements personally ... would @J535D165 @holub008 mind giving this a review too once the tests pass to make sure the changes look appropriate?

File: rispy/parser.py
Comment (J535D165) [reply rispy/parser.py:53]: I agree with this. One note: if `list_tags` is set to an empty list (because you don't want list items), this will result in unexpected behavior, isn't it.

File: rispy/parser.py
Comment (J535D165) [reply rispy/parser.py:26]: And make this an ABC?

File: rispy/parser.py
Comment (J535D165) [inline rispy/parser.py:200]: `@abstractmethod`

File: rispy/parser.py
Comment (shapiromatron) [reply rispy/parser.py:53]: good catch!

**Comment by J535D165** [comment]
I'm still not sure about passing an object instead of a class to load/loads/dump/dumps. The rispy API is inspired by the `json` API in CPython. That API accepts only classes (the argument is named `cls`). See implementation: https://github.com/python/cpython/blob/021169f7685bcc733f8c23985c5140d483021f60/Lib/json/__init__.py#L299-L359. I think it is more pythonic to do things that way.

**Comment by J535D165** [comment]
Btw, I'm looking forward to sharing the custom implementations (next to the 2 official standards). What is your idea about sharing those? A contrib folder? Or just include them in the official repo? 

Example usage: 
```
from rispy.contrib.implementation import RayyanRIS, EndNoteRIS
```

In that case, what do we do with WOK?

Might be a bit off-topic, however, it might be relevant for some parts of this PR.

**Comment by scott-8** [comment]
The reason I prefer passing an object is it makes it very easy to use custom tags, since you can just create an instance. If we went with passing a custom class, I think it would be good to make sure that this is still simple. To me, creating a new class just to change the name of a tag doesn't seem like the way to go, but I'm open to input.

Regarding custom implementations, I think some sort of contrib folder would be a good idea to avoid being overwhelmed by new implementations. I think it would be okay to move WOK to that folder, since the only way to use it now is to import it separately anyways.

**Comment by J535D165** [comment]
Just to be sure, I love your work on this PR. As one of the maintainers of https://github.com/asreview/asreview, I super grateful to depend on this package and your contributions. However, it's important to have stable dependencies. So if a good discussion about the API can avoid future API changes, that would be helpful. 

I think the `json` API is a very flexible implementation. Mainly because additional keyword arguments are passed to the constructor of the class. Therefore, you don't have to define a new object. 

The implementation in this PR is as follows (correct me if I'm wrong): 
```
parser = rispy.RisParser(list_tags=["AU", "SN"])
rispy.loads(fp, implementation=parser)
```

In line with the JSON API, you would write:
```
rispy.loads(fp, implementation=rispy.RisParser, list_tags=["AU", "SN"])
```
Because this is a simple change to the implementation, you only have to pass the keyword arguments to the constructor of the existing class. For more complex implementations, like WOK, you need to define a new class. But this isn't different from the current implementation. 

Because `rispy.RisParser` is the default parser, you can simplify this code:
```
rispy.loads(fp, list_tags=["AU", "SN"])
```

Things are now super flexible and extensible. I think this is all we need. 

Please let me know what you think. I think this only requires minor changes to your PR :) Mainly the constructors of the classes.

**Comment by scott-8** [comment]
I think you make a very good point. Originally I was looking at your comment above and assuming you meant a change like this:

```python
from rispy import RisParser

class MyCustomRisParser(RisParser):
   list_tags = ["KW", "N1"]
```

but the idea you presented seems like a much better way of doing it. I'd be okay with this change, and it would not be hard to implement. I'll let @shapiromatron weigh in on this before I make changes.

**Comment by shapiromatron** [comment]
> The reason I prefer passing an object is it makes it very easy to use custom tags, since you can just create an instance. If we went with passing a custom class, I think it would be good to make sure that this is still simple. To me, creating a new class just to change the name of a tag doesn't seem like the way to go, but I'm open to input.
> 
> Regarding custom implementations, I think some sort of contrib folder would be a good idea to avoid being overwhelmed by new implementations. I think it would be okay to move WOK to that folder, since the only way to use it now is to import it separately anyways.

Yep, creating `contrib.implementations` seems simple and easy to maintain and accept new PR. I agree, moving the WOK there makes sense b/c AFAIK it's infrequently used...

**Comment by shapiromatron** [comment]
> I think you make a very good point. Originally I was looking at your comment above and assuming you meant a change like this:
> 
> ```python
> from rispy import RisParser
> 
> class MyCustomRisParser(RisParser):
>    list_tags = ["KW", "N1"]
> ```
> 
> but the idea you presented seems like a much better way of doing it. I'd be okay with this change, and it would not be hard to implement. I'll let @shapiromatron weigh in on this before I make changes.

Thanks for letting me catch up @scott-8. I like @J535D165 suggestion of accepting classes for implementation and then any **kw are passed to the constructor of the implementation. I actually didn't know that json.loads/dumps does this, but the closer the we stay to that API the more easy to understand this package is IMO. https://docs.python.org/3/library/json.html#basic-usage

**Comment by J535D165** [comment]
Nice update @scott-8! It's also good to see that there are fewer breaking changes now. 

I'm not sure about the class variables. I think we should move them to the constructor (at least the most-used ones, but maybe all of them). There is no need to have class variables anymore because it is easy to overwrite. Custom implementations can overwrite the constructor (limited number of use-cases now). 

> > The reason I prefer passing an object is it makes it very easy to use custom tags, since you can just create an instance. If we went with passing a custom class, I think it would be good to make sure that this is still simple. To me, creating a new class just to change the name of a tag doesn't seem like the way to go, but I'm open to input.
> > Regarding custom implementations, I think some sort of contrib folder would be a good idea to avoid being overwhelmed by new implementations. I think it would be okay to move WOK to that folder, since the only way to use it now is to import it separately anyways.
> 
> Yep, creating `contrib.implementations` seems simple and easy to maintain and accept new PR. I agree, moving the WOK there makes sense b/c AFAIK it's infrequently used...

I'm happy to implement this after this PR is merged.

**Comment by scott-8** [comment]
Thanks for the comments @J535D165. I agree that some of them could be moved to the constructor, but I'm not sure that all of them should be. One suggestion would be to move all of the booleans (e.g. `SKIP_MISSING_TAGS`, `SKIP_UNKNOWN_TAGS`, etc.) but keep the other class variables. I like the idea of keeping some of the parameters as class variables to facilitate the creation of custom implementations. If you are dealing with a non-standard RIS format that needs a custom parser, it's much simpler and cleaner to override class variables for the user than overwriting the constructor or feeding every parameter to `load` on each call. I don't mind too much either way, however.

File: rispy/parser.py
Comment (shapiromatron) [inline rispy/parser.py:92]: @J535D165 was correct here; if list_tags is falsy, like [], then this will use the default, even if you intentionally specified no skip tags - perhaps change to 

```python
self.mapping = mapping if mapping is not None else self. DEFAULT_MAPPING
self.list_tags = list_tags if list_tags is not None else self. DEFAULT_LIST_TAGS
```

File: README.rst
Comment (shapiromatron) [inline README.rst:258]: suggested edit: A custom writer class can inherit from  ``BaseWriter`` or one of it's subclasses such as ``RisWriter``.

File: rispy/parser.py
Comment (shapiromatron) [inline rispy/parser.py]: change `IGNORE` to `DEFAULT_IGNORE` and add optional init argument for ignore?

File: rispy/writer.py
Comment (shapiromatron) [inline rispy/writer.py:20]: this should be an ABC too

**Review by shapiromatron — CHANGES_REQUESTED** [review]
few very minor edits

**Review by shapiromatron — APPROVED** [review]
Nice job, this seems like a great addition!

**Review by J535D165 — APPROVED** [review]

File: rispy/writer.py
Comment (J535D165) [inline rispy/writer.py:188]: `**kw` arguments should also be passed.

**Review by J535D165 — CHANGES_REQUESTED** [review]

File: rispy/writer.py
Comment (J535D165) [reply rispy/writer.py:188]: Apparently, this is not covered by a test.

Comment (scott-8) [reply rispy/writer.py:188]: nice catch, I'll add a test case as well

Comment (J535D165) [reply rispy/writer.py:188]: Great

**Comment by J535D165** [comment]
Great work @scott-8

**Comment by scott-8** [comment]
Thanks for the help everyone!

**Pull request merged**
